### DISCUSSION TITLE
Number of instruction miscalculated - Chapter5 Section4

### DISCUSSION THREAD

**Discussion by bernabeu-a** [goossens-springer/goossens-book-ip-projects#4] [Show and tell]
Hi,

First, thank you for your work and for this wonderful book. I'm using the Alchitry AU FPGA to run the design presented in the book.
I encountered the following behavior:
- In the chapter 5, section 4 when we add the decoder, while in the simulation all worked fine with the 14 instructions fetched and decoded from the test_op_imm source code, when receiving data from the FPGA, only 1 instruction was displayed as fetched and decoded.
- To make sure that all the 14 instructions was fetched and decoded, I added a variable with the last value of the **PC** and it was 14 as expected. Thus the problem come from the _nbi_ variable and its computation.
- So it look like that the HLS compiler might optimize (too much) the design presented in this section. 
- Adding the **static** keyword on the _nbi_ variable in the HLS design seems to fix this issue.

Regards,

 Antoine Bernabeu

**Comment by goossens-springer** [comment]
Hello Antoine, 

Nice. I take note of that. 

Are you from Polytech Nantes ? 

bg. 


De: "bernabeu-a" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Subscribed" ***@***.***> 
Envoyé: Vendredi 24 Mai 2024 11:18:50 
Objet: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Hi, 

First, thank you for your work and for this wonderful book. I'm using the Alchitry AU FPGA to run the design presented in the book. 
I encountered the following behavior: 

    * In the chapter 5, section 4 when we add the decoder, while in the simulation all worked fine with the 14 instructions fetched and decoded from the test_op_imm source code, when receiving data from the FPGA, only 1 instruction was displayed as fetched and decoded. 
    * To make sure that all the 14 instructions was fetched and decoded, I added a variable with the last value of the PC and it was 14 as expected. Thus the problem come from the nbi variable and its computation. 
    * So it look like that the HLS compiler might optimize (too much) the design presented in this section. 
    * Adding the static keyword on the nbi variable in the HLS design seems to fix this issue. 


Regards, 

Antoine Bernabeu

**Reply by bernabeu-a** [reply]
Hi,

Yes I am actually working at IETR in Polytech Nantes.

Antoine

**Comment by goossens-springer** [comment]
Hello Antoine, 

As far as I have seen, the Alchitry AU FPGA is the XC7A35T-1C, which is the same FPGA as the one used on the Basys3 board. 
I have tested all the codes given in the book os the Basys3 board. 
In the github Readme file, there are many useful informations concerning the Basys3 board (to be added to the ones given in the book). 
As far as I remember, I had no problem with the chapter 5 experiments. 
Maybe there are differences in the system on the FPGA itself. 

Please, would you be kind enough to send me all the problems you might encounter on the Alchitry AU board. 
I would add a special section for this board on the github Readme file for later users. 

Anyway, thank you for your feedback. 
bg. 


De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Envoyé: Vendredi 24 Mai 2024 12:06:21 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello Antoine, 

Nice. I take note of that. 

Are you from Polytech Nantes ? 

bg. 


De: "bernabeu-a" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Subscribed" ***@***.***> 
Envoyé: Vendredi 24 Mai 2024 11:18:50 
Objet: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Hi, 

First, thank you for your work and for this wonderful book. I'm using the Alchitry AU FPGA to run the design presented in the book. 
I encountered the following behavior: 

    * In the chapter 5, section 4 when we add the decoder, while in the simulation all worked fine with the 14 instructions fetched and decoded from the test_op_imm source code, when receiving data from the FPGA, only 1 instruction was displayed as fetched and decoded. 
    * To make sure that all the 14 instructions was fetched and decoded, I added a variable with the last value of the PC and it was 14 as expected. Thus the problem come from the nbi variable and its computation. 
    * So it look like that the HLS compiler might optimize (too much) the design presented in this section. 
    * Adding the static keyword on the nbi variable in the HLS design seems to fix this issue. 


Regards, 

Antoine Bernabeu

**Reply by bernabeu-a** [reply]
Hi,

Sure, for documentation, here are the differences I've noticed so far to use the Alchitry AU FPGA with the materials of the book. 

- When creating Block Design in Vivado, we must add a uartlite block and connect the TX uart port of the block to an output port labelled with the right port from the constraint file. Assuming the constraint file provided by Alchitry, a port named **usb_tx**.
- For the input clock, same procedure, we must add an input port with the right label from the constraint file. Assuming the constraint file provided by Alchitry, a port named **clk**.
- For the reset, same procedure.  Assuming the constraint file provided by Alchitry, a port named **rst_n**.

As example, I add some screenshot.
![rst_clk](https://github.com/goossens-springer/goossens-book-ip-projects/assets/99437435/9bd69716-3c10-4772-a2e8-fa73b3b901ce)
![uartlite](https://github.com/goossens-springer/goossens-book-ip-projects/assets/99437435/6560d00d-3009-4332-874c-7aa6ba98f9b0)

Regarding loading the IP with Vitis:
- We must use xil_printf instead of printf to not exceed memory usage.
- To flash to the FPGA, once the project is compiled, click on the Vitis menu, then program device. In the Software Configuration part, click on bootloop and choose browse. Find the elf file (in workspace/application_name/Debug/). Then click generate. This will create a bitstream file that we can load using the alchitry-loader software. The bitstream file is in workspace/application_name/_ide/bitstream.

That almost all for the moment.

Antoine

**Comment by goossens-springer** [comment]
I know very well Sébastien Pillement in Nantes. Do you know him ? 

bg. 


De: "bernabeu-a" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Vendredi 31 Mai 2024 10:22:49 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Hi, 

Yes I am actually working at IETR in Polytech Nantes. 

Antoine

**Reply by bernabeu-a** [reply]
Yes I'm actually in a post-doc position under his supervision :)

ab

**Comment by goossens-springer** [comment]
forward him my regards !! (did he suggest you the book or did you find it by yourself ?) 


De: "bernabeu-a" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Vendredi 31 Mai 2024 10:41:57 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Yes I'm actually in a post-doc position under his supervision :) 

ab

**Comment by goossens-springer** [comment]
Ok, I will add these Alchetry adaptations to the Readme file. Is it ok for you if I mention your name and affiliation in the readme file ? 


De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Envoyé: Vendredi 31 Mai 2024 11:08:53 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

forward him my regards !! (did he suggest you the book or did you find it by yourself ?) 


De: "bernabeu-a" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Vendredi 31 Mai 2024 10:41:57 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Yes I'm actually in a post-doc position under his supervision :) 

ab

**Reply by bernabeu-a** [reply]
Sure, you can mention my name and affiliation.

One last thing, for vitis_hls and vivado, the parts used should be xc7a35tftg256-1

Regards,

Antoine

**Comment by atreistm** [comment]
Thank you for the detailed response. I opened up a new issue, but I now see that it is related to this one.
I am using Vivado and Vitis 2024.1 and an Arty-Z7-20. The first project went well. However , for the fetching_decode_ip and fde_ip projects, I am having the same problem with nbi. Could you please detail which file you changed so that I can try it out? There may be other differences/bugs introduced by Vitis 2024.1 and I am trying to weed them out. 
I purchased the fascinating book as I am in the process of moving our undergraduate Computer Architecture course from MIPS to RISC-V. The course is for computer science students who have no knowledge of FPGAs. At this point, my study is to get a better understanding of how the CPU is implemented so enhance my lectures.

**Comment by goossens-springer** [comment]
Hello Avraham, 

The Arty-Z7 20 has the same FPGA than my Pynq-Z2 (XC7Z020), so you should be able to run all the experiments as they are presented in the main parts of the book. 

However, I have not tested anything after the 2023.2 version of the Vitis suite. I will try with 2024.1 but I need some time (to make room on my hard drive to install the greedy Vitis software). 

I did not change anything in my files up to now. 

Keep in touch, 

Bernard. 
PS : which university are you from ? What is your lecture about ? If you wish, we can discuss about that. 


De: "Avraham Treistman" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Vendredi 16 Août 2024 16:01:23 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Thank you for the detailed response. I opened up a new issue, but I now see that it is related to this one. 
I am using Vivado and Vitis 2024.1 and an Arty-Z7-20. The first project went well. However , for the fetching_decode_ip and fde_ip projects, I am having the same problem with nbi. Could you please detail which file you changed so that I can try it out? There may be other differences/bugs introduced by Vitis 2024.1 and I am trying to weed them out. 
I purchased the fascinating book as I am in the process of moving our undergraduate Computer Architecture course from MIPS to RISC-V. The course is for computer science students who have no knowledge of FPGAs. At this point, my study is to get a better understanding of how the CPU is implemented so enhance my lectures.

**Comment by atreistm** [comment]
Hi,
Thank you for your prompt response. I will send a more detailed response in the next day or two. I tried the same steps on my 2023 version and I have the same issue. On GitHub, someone suggested making NBI a static variable. This did not work for me, but maybe I changed it in the wrong place. There is a small fix I made in the helloworld.c file. The lookup function takes a baseaddr and not a device ID. I will send you the code.
More to come.

All the best,
Avi

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by bernabeu-a** [reply]
Hi,

I only added the static when first defining the variable, in the main ip (fde_ip or fetch_decode_ip). I used the latest version of vitis hls.
<img width="592" alt="Capture d’écran 2024-08-18 à 09 27 44" src="https://github.com/user-attachments/assets/a3dab3f9-5760-4b5c-b9e9-0fe4275385b4">

Hope, it will help you.

Regards,

Antoine

**Comment by atreistm** [comment]
Hi,

I tried this change and I am still getting an output of 1.

1 fetched, decoded and executed instructions

In the code of the Vitis program, the only change I had to make was
  cfg_ptr = XFde_ip_LookupConfig(XPAR_XFDE_IP_0_BASEADDR);

This change was introduced in version 2023.2, where DEVICE_O was dropped.

I am not sure what to change at this point to get the program to work, but would be happy to try to figure this one out.
You mentioned that you last tested on 2023.2. Did you run the classic versions or the unified version?

I will try to the classic versions on Tuesday, see if anything improves.

Thanks,
Avi

**Reply by bernabeu-a** [reply]
Oh my bad, I thought the problem was with the simulator, in vitis HLS. 
For the implementation, I only used the classic version of vitis. 

Regards,
Antoine

**Comment by goossens-springer** [comment]
Hello all, 

I will check on my side this evening after work and let you know. (on 2023.1 ; I'll install 2024.1 in the night and check tomorrow). 

Bernard. 


De: "bernabeu-a" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Lundi 19 Août 2024 08:11:23 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 





Oh my bad, I thought the problem was with the simulator, in vitis HLS. 
For the implementation, I only used the classic version of vitis. 

Regards, 
Antoine

**Comment by atreistm** [comment]
I did another experiment last night. In 2023.2 classic I opened the original files from GitHub and launched them without recompiling the vivado or.hls. I only compiled the vitis app and it ran correctly. However, as oon as I recompiled

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by atreistm** [comment]
However, as soon as I recompiled everything, it stopped working.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by goossens-springer** [comment]
Hello All, 

As the 2023 version of Vitis makes some major changes, I'll redo the full experiments and add a 2023.2 folder to the repo. It will take some time though. Be patient (moreover, I do have some holidays this year !) 

Bernard. 


De: "Avraham Treistman" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Lundi 19 Août 2024 08:59:34 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 



However, as soon as I recompiled everything, it stopped working. 

Get Outlook for Android<https://aka.ms/AAb9ysg> 
________________________________

**Comment by atreistm** [comment]
Dear Bernard,

Thank you for your assistance. I am very curious as to what changed in Vivado/Vitis. I am still trying to figure it out. I will of course update you if I find anything.

I teach at the Lev Academic Center in Jerusalem. Our college focuses primarily on electronic and computer engineering. We are not as well known as the Technion, but we are one of Israel's most prominent tech schools. The course I have been teaching for the last twenty years is 1st year Computer Architecture based on Patterson and Hennessey using MIPS. I have decided to move the course to RISC-V. My lecturers are all in a "panic", I am trying my best to calm them down. One of the most significant changes I am introducing is in the labs. We have always used the MARS emulator. I have purchased an ESP32C6 development board for each student. I hope it will bring their assembly programming skills to a more practical level.

Sometime in the future, I would like to introduce them to FPGAs. I teach only computer science students. Unfortunately, they don't get too much hands-on experience with hardware. I hope to change this in the following years. I would love to offer a course in FPGA, but it might be a while before I can get approval.

My primary research for the last few years has been in text mining, and It will be interesting to track the convergence of microcontrollers and AI.

Enjoy your holidays!

Looking forward to your updates,

Avi

**Comment by goossens-springer** [comment]
Hello Avi, 

I am working on the new Vitis now. I have found that on the HLS side, you can import an HLS project previously developed (i.e. the ones given on the github). I have also found that you don't even need to do that if you just want to rebuild the design under Vivado. Then, you just need to update the IPs in the design (because they have been upgraded in the new Vivado library). 

So, I will add a few indications on how to redo the experiments on Vitis/Vivado starting from 2023.1 in the readme.md file. 

I will also check the problem with the fetching_decoding_ip (nbi not correct). 

Concerning your teaching duty, you can feel yourself happy as you will be able to teach REAL hardware (through an FPGA simulation). As I say in the foreword of the book, a ll my life 
I have been teaching theoretical architecture to my students with simulators, but no realization. Having a home made processor running a code is really a great sensation/feeling ! 

Moreover, with HLS, software people can tackle hardware with quite no effort as they just code in C. 

To convince your colleagues about FPGA/architecture/processors/assembly language/ISA/RISC-V, tell them that everything is linked. Not knowing hardware means programming dumbly, with no idea on how to save space (how data is represented in memory, what is a stack, a heap, global data ...) and how to save time in low level functions (which are the most used ones). In order to program correctly, you need to know : 

- algorithmic, 
- data structures and programming methods (like sorting, searching and so on), 
- a programming language (C for low level, C++ for higher level), 
- compiling (what is the code after the translation, i.e. assembly/ISA), 
- architecture (ISA), 
- OS (the code uses computing resources, memory and I/O resources ; how are they handled by the OS ?) 

These are the basics. 

And to teach anything means teaching how to read and write in a given language. For architecture, it means being able to read a design and to build one. With HLS, it is simply a matter of reading a C program and write one, with a particular architectural point of view. 

But you are far from being a newbie and you surely have your own ideas on that (however, as somebody has said once -and I loved the sentence- "I have been young longer than you"). 

I wish you the best Avi (particularly in these difficult times) and feel free to interact. By the way, if you go on my "very old style" web page, you can find my email address. 

Bernard. 


De: "Avraham Treistman" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Mardi 20 Août 2024 10:16:18 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 



Dear Bernard, 

Thank you for your assistance. I am very curious as to what changed in Vivado/Vitis. I am still trying to figure it out. I will of course update you if I find anything. 

I teach at the Lev Academic Center in Jerusalem. Our college focuses primarily on electronic and computer engineering. We are not as well known as the Technion, but we are one of Israel's most prominent tech schools. The course I have been teaching for the last twenty years is 1st year Computer Architecture based on Patterson and Hennessey using MIPS. I have decided to move the course to RISC-V. My lecturers are all in a "panic", I am trying my best to calm them down. One of the most significant changes I am introducing is in the labs. We have always used the MARS emulator. I have purchased an ESP32C6 development board for each student. I hope it will bring their assembly programming skills to a more practical level. 

Sometime in the future, I would like to introduce them to FPGAs. I teach only computer science students. Unfortunately, they don't get too much hands-on experience with hardware. I hope to change this in the following years. I would love to offer a course in FPGA, but it might be a while before I can get approval. 

My primary research for the last few years has been in text mining, and It will be interesting to track the convergence of microcontrollers and AI. 

Enjoy your holidays! 

Looking forward to your updates, 

Avi

**Comment by goossens-springer** [comment]
Hello all, 

I have run the fetching_decoding_ip on the vitis 2023.2 distro. As Antoine and Avi have pointed out, the printed "nbi" is 1 instead of 14. 

I'll have to investigate the hls source code to understand what went wrong (the simulation printed an "nbi" of 14, as it should be). 

More on that issue later. 

Bernard. 


De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Envoyé: Mardi 20 Août 2024 11:32:53 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello Avi, 

I am working on the new Vitis now. I have found that on the HLS side, you can import an HLS project previously developed (i.e. the ones given on the github). I have also found that you don't even need to do that if you just want to rebuild the design under Vivado. Then, you just need to update the IPs in the design (because they have been upgraded in the new Vivado library). 

So, I will add a few indications on how to redo the experiments on Vitis/Vivado starting from 2023.1 in the readme.md file. 

I will also check the problem with the fetching_decoding_ip (nbi not correct). 

Concerning your teaching duty, you can feel yourself happy as you will be able to teach REAL hardware (through an FPGA simulation). As I say in the foreword of the book, a ll my life 
I have been teaching theoretical architecture to my students with simulators, but no realization. Having a home made processor running a code is really a great sensation/feeling ! 

Moreover, with HLS, software people can tackle hardware with quite no effort as they just code in C. 

To convince your colleagues about FPGA/architecture/processors/assembly language/ISA/RISC-V, tell them that everything is linked. Not knowing hardware means programming dumbly, with no idea on how to save space (how data is represented in memory, what is a stack, a heap, global data ...) and how to save time in low level functions (which are the most used ones). In order to program correctly, you need to know : 

- algorithmic, 
- data structures and programming methods (like sorting, searching and so on), 
- a programming language (C for low level, C++ for higher level), 
- compiling (what is the code after the translation, i.e. assembly/ISA), 
- architecture (ISA), 
- OS (the code uses computing resources, memory and I/O resources ; how are they handled by the OS ?) 

These are the basics. 

And to teach anything means teaching how to read and write in a given language. For architecture, it means being able to read a design and to build one. With HLS, it is simply a matter of reading a C program and write one, with a particular architectural point of view. 

But you are far from being a newbie and you surely have your own ideas on that (however, as somebody has said once -and I loved the sentence- "I have been young longer than you"). 

I wish you the best Avi (particularly in these difficult times) and feel free to interact. By the way, if you go on my "very old style" web page, you can find my email address. 

Bernard. 


De: "Avraham Treistman" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Mardi 20 Août 2024 10:16:18 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 



Dear Bernard, 

Thank you for your assistance. I am very curious as to what changed in Vivado/Vitis. I am still trying to figure it out. I will of course update you if I find anything. 

I teach at the Lev Academic Center in Jerusalem. Our college focuses primarily on electronic and computer engineering. We are not as well known as the Technion, but we are one of Israel's most prominent tech schools. The course I have been teaching for the last twenty years is 1st year Computer Architecture based on Patterson and Hennessey using MIPS. I have decided to move the course to RISC-V. My lecturers are all in a "panic", I am trying my best to calm them down. One of the most significant changes I am introducing is in the labs. We have always used the MARS emulator. I have purchased an ESP32C6 development board for each student. I hope it will bring their assembly programming skills to a more practical level. 

Sometime in the future, I would like to introduce them to FPGAs. I teach only computer science students. Unfortunately, they don't get too much hands-on experience with hardware. I hope to change this in the following years. I would love to offer a course in FPGA, but it might be a while before I can get approval. 

My primary research for the last few years has been in text mining, and It will be interesting to track the convergence of microcontrollers and AI. 

Enjoy your holidays! 

Looking forward to your updates, 

Avi

**Comment by goossens-springer** [comment]
Hello all, 

I succeeded to run the fetching_decoding_ip on the vitis 2023.2 distro. (printing "14 fetched and decoded instructions") 

I did not change anything except ... I have fully rebuilt the design in Vivado (from a clean diagram, with new components from the library and the new version of the fetching_decoding_ip. 
Then, I have regenerated the bitstream. 
So, the lesson from this is that you have to rebuild everything. No changing to the code, but rebuild the exported IP from the HLS design, rebuild the design in Vivado, regenerate the bitstream, re-export the hardware and rebuild the final platform. 

Please let me know if this works for you. 

Bernard. 

De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Envoyé: Mardi 20 Août 2024 19:25:57 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello all, 

I have run the fetching_decoding_ip on the vitis 2023.2 distro. As Antoine and Avi have pointed out, the printed "nbi" is 1 instead of 14. 

I'll have to investigate the hls source code to understand what went wrong (the simulation printed an "nbi" of 14, as it should be). 

More on that issue later. 

Bernard. 


De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Envoyé: Mardi 20 Août 2024 11:32:53 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello Avi, 

I am working on the new Vitis now. I have found that on the HLS side, you can import an HLS project previously developed (i.e. the ones given on the github). I have also found that you don't even need to do that if you just want to rebuild the design under Vivado. Then, you just need to update the IPs in the design (because they have been upgraded in the new Vivado library). 

So, I will add a few indications on how to redo the experiments on Vitis/Vivado starting from 2023.1 in the readme.md file. 

I will also check the problem with the fetching_decoding_ip (nbi not correct). 

Concerning your teaching duty, you can feel yourself happy as you will be able to teach REAL hardware (through an FPGA simulation). As I say in the foreword of the book, a ll my life 
I have been teaching theoretical architecture to my students with simulators, but no realization. Having a home made processor running a code is really a great sensation/feeling ! 

Moreover, with HLS, software people can tackle hardware with quite no effort as they just code in C. 

To convince your colleagues about FPGA/architecture/processors/assembly language/ISA/RISC-V, tell them that everything is linked. Not knowing hardware means programming dumbly, with no idea on how to save space (how data is represented in memory, what is a stack, a heap, global data ...) and how to save time in low level functions (which are the most used ones). In order to program correctly, you need to know : 

- algorithmic, 
- data structures and programming methods (like sorting, searching and so on), 
- a programming language (C for low level, C++ for higher level), 
- compiling (what is the code after the translation, i.e. assembly/ISA), 
- architecture (ISA), 
- OS (the code uses computing resources, memory and I/O resources ; how are they handled by the OS ?) 

These are the basics. 

And to teach anything means teaching how to read and write in a given language. For architecture, it means being able to read a design and to build one. With HLS, it is simply a matter of reading a C program and write one, with a particular architectural point of view. 

But you are far from being a newbie and you surely have your own ideas on that (however, as somebody has said once -and I loved the sentence- "I have been young longer than you"). 

I wish you the best Avi (particularly in these difficult times) and feel free to interact. By the way, if you go on my "very old style" web page, you can find my email address. 

Bernard. 


De: "Avraham Treistman" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Mardi 20 Août 2024 10:16:18 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 



Dear Bernard, 

Thank you for your assistance. I am very curious as to what changed in Vivado/Vitis. I am still trying to figure it out. I will of course update you if I find anything. 

I teach at the Lev Academic Center in Jerusalem. Our college focuses primarily on electronic and computer engineering. We are not as well known as the Technion, but we are one of Israel's most prominent tech schools. The course I have been teaching for the last twenty years is 1st year Computer Architecture based on Patterson and Hennessey using MIPS. I have decided to move the course to RISC-V. My lecturers are all in a "panic", I am trying my best to calm them down. One of the most significant changes I am introducing is in the labs. We have always used the MARS emulator. I have purchased an ESP32C6 development board for each student. I hope it will bring their assembly programming skills to a more practical level. 

Sometime in the future, I would like to introduce them to FPGAs. I teach only computer science students. Unfortunately, they don't get too much hands-on experience with hardware. I hope to change this in the following years. I would love to offer a course in FPGA, but it might be a while before I can get approval. 

My primary research for the last few years has been in text mining, and It will be interesting to track the convergence of microcontrollers and AI. 

Enjoy your holidays! 

Looking forward to your updates, 

Avi

**Comment by goossens-springer** [comment]
Hello all, 

Sorry, I was wrong. You have to change the "fetching_decoding_ip.cpp" file the following way (eliminate the "nbi" intermediate variable and directly compute in the nb_instruction variable ; this is probably an inadequate optimisation of the synthesizer): 

void fetching_decoding_ip ( 
unsigned int start_pc , 
unsigned int code_ram [CODE_RAM_SIZE], 
unsigned int * nb_instruction ){ 
#pragma HLS INTERFACE s_axilite port = start_pc 
#pragma HLS INTERFACE s_axilite port = code_ram 
#pragma HLS INTERFACE s_axilite port = nb_instruction 
#pragma HLS INTERFACE s_axilite port = return 
code_address_t pc; 
instruction_t instruction; 
bit_t is_running; 
decoded_instruction_t d_i; 
pc = start_pc; 
*nb_instruction = 0 ; 
do { 
#pragma HLS PIPELINE II =3 
fetch (pc, code_ram, &instruction); 
decode (instruction, &d_i); 
execute (pc, d_i, &pc); 
statistic_update (nb_instruction); 
running_cond_update (instruction, &is_running); 
} while (is_running); 
} 

Bernard. 


De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Envoyé: Mercredi 21 Août 2024 00:11:32 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello all, 

I succeeded to run the fetching_decoding_ip on the vitis 2023.2 distro. (printing "14 fetched and decoded instructions") 

I did not change anything except ... I have fully rebuilt the design in Vivado (from a clean diagram, with new components from the library and the new version of the fetching_decoding_ip. 
Then, I have regenerated the bitstream. 
So, the lesson from this is that you have to rebuild everything. No changing to the code, but rebuild the exported IP from the HLS design, rebuild the design in Vivado, regenerate the bitstream, re-export the hardware and rebuild the final platform. 

Please let me know if this works for you. 

Bernard. 

De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Envoyé: Mardi 20 Août 2024 19:25:57 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello all, 

I have run the fetching_decoding_ip on the vitis 2023.2 distro. As Antoine and Avi have pointed out, the printed "nbi" is 1 instead of 14. 

I'll have to investigate the hls source code to understand what went wrong (the simulation printed an "nbi" of 14, as it should be). 

More on that issue later. 

Bernard. 


De: "Bernard Goossens" ***@***.***> 
À: "goossens-springer/goossens-book-ip-projects" ***@***.***> 
Envoyé: Mardi 20 Août 2024 11:32:53 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 

Hello Avi, 

I am working on the new Vitis now. I have found that on the HLS side, you can import an HLS project previously developed (i.e. the ones given on the github). I have also found that you don't even need to do that if you just want to rebuild the design under Vivado. Then, you just need to update the IPs in the design (because they have been upgraded in the new Vivado library). 

So, I will add a few indications on how to redo the experiments on Vitis/Vivado starting from 2023.1 in the readme.md file. 

I will also check the problem with the fetching_decoding_ip (nbi not correct). 

Concerning your teaching duty, you can feel yourself happy as you will be able to teach REAL hardware (through an FPGA simulation). As I say in the foreword of the book, a ll my life 
I have been teaching theoretical architecture to my students with simulators, but no realization. Having a home made processor running a code is really a great sensation/feeling ! 

Moreover, with HLS, software people can tackle hardware with quite no effort as they just code in C. 

To convince your colleagues about FPGA/architecture/processors/assembly language/ISA/RISC-V, tell them that everything is linked. Not knowing hardware means programming dumbly, with no idea on how to save space (how data is represented in memory, what is a stack, a heap, global data ...) and how to save time in low level functions (which are the most used ones). In order to program correctly, you need to know : 

- algorithmic, 
- data structures and programming methods (like sorting, searching and so on), 
- a programming language (C for low level, C++ for higher level), 
- compiling (what is the code after the translation, i.e. assembly/ISA), 
- architecture (ISA), 
- OS (the code uses computing resources, memory and I/O resources ; how are they handled by the OS ?) 

These are the basics. 

And to teach anything means teaching how to read and write in a given language. For architecture, it means being able to read a design and to build one. With HLS, it is simply a matter of reading a C program and write one, with a particular architectural point of view. 

But you are far from being a newbie and you surely have your own ideas on that (however, as somebody has said once -and I loved the sentence- "I have been young longer than you"). 

I wish you the best Avi (particularly in these difficult times) and feel free to interact. By the way, if you go on my "very old style" web page, you can find my email address. 

Bernard. 


De: "Avraham Treistman" ***@***.***> 
À: "goossens-springer" ***@***.***> 
Cc: "Bernard Goossens" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Mardi 20 Août 2024 10:16:18 
Objet: Re: [goossens-springer/goossens-book-ip-projects] Number of instruction miscalculated - Chapter5 Section4 (Discussion #4) 



Dear Bernard, 

Thank you for your assistance. I am very curious as to what changed in Vivado/Vitis. I am still trying to figure it out. I will of course update you if I find anything. 

I teach at the Lev Academic Center in Jerusalem. Our college focuses primarily on electronic and computer engineering. We are not as well known as the Technion, but we are one of Israel's most prominent tech schools. The course I have been teaching for the last twenty years is 1st year Computer Architecture based on Patterson and Hennessey using MIPS. I have decided to move the course to RISC-V. My lecturers are all in a "panic", I am trying my best to calm them down. One of the most significant changes I am introducing is in the labs. We have always used the MARS emulator. I have purchased an ESP32C6 development board for each student. I hope it will bring their assembly programming skills to a more practical level. 

Sometime in the future, I would like to introduce them to FPGAs. I teach only computer science students. Unfortunately, they don't get too much hands-on experience with hardware. I hope to change this in the following years. I would love to offer a course in FPGA, but it might be a while before I can get approval. 

My primary research for the last few years has been in text mining, and It will be interesting to track the convergence of microcontrollers and AI. 

Enjoy your holidays! 

Looking forward to your updates, 

Avi
